### PR TITLE
Make use of more Higher Order Types

### PR DESCRIPTION
**Pull request by Cypher1**
Working on #3030 replacing some of our ad-hoc indexed types with Dictionary<T>s.

### CODE REVIEW

File: src/runtime/storage/storage-provider-base.ts [hunk 1]
```diff
@@ -15,14 +15,14 @@ import {ModelValue, SerializedModelEntry} from './crdt-collection-model.js';
...
-
+import {Dictionary, Consumer} from '../hot.js';
 
 enum EventKind {
   change = 'Change'
 }
 
 // tslint:disable-next-line: no-any
-type Callback = (v: {[index: string]: any}) => void;
+type Callback = Consumer<Dictionary<any>>;
```
Comment (alxrsngrtn) [inline src/runtime/storage/storage-provider-base.ts]: Can we namespace this callback further? Like a `KeywordArgumentCallback` (`KwargCallback` for short).

File: src/runtime/services.ts [hunk 2]
```diff
@@ -9,9 +9,7 @@
 
 import {Dictionary} from './hot.js';
 
-export interface Service {
-  [name: string]: Function;
-}
+export type Service = Dictionary<Function>;
```
[the remark this answers is not in the document]
Comment (alxrsngrtn) [reply src/runtime/services.ts]: On second though... I think this was intentional. 

I want to create another name for services, not merely create an alias. 

https://www.typescriptlang.org/docs/handbook/advanced-types.html#interfaces-vs-type-aliases

This doesn't matter much for now. However, I expect `Service` will change back to an Interface in the (near) future.

File: src/runtime/recipe/comparable.ts [hunk 3]
```diff
@@ -38,7 +39,7 @@ export function compareArrays<a>(a1: a[], a2: a[], compare: (first: a, second: a
   return 0;
 }
 
-export function compareObjects<a>(o1: {[index: string]: a} | null, o2: {[index: string]: a} | null, compare: (first: a, second: a) => number) {
+export function compareObjects<a>(o1: Dictionary<a> | null, o2: Dictionary<a> | null, compare: (first: a, second: a) => number) {
```
[the remark this answers is not in the document]
Comment (alxrsngrtn) [reply src/runtime/recipe/comparable.ts]: Could they be `Record` types? https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types

File: src/planning/strategizer.ts [hunk 4]
```diff
@@ -11,20 +11,21 @@ import {Recipe} from '../runtime/recipe/recipe.js';
...
   generatedDerivations?: number;
   nullDerivations?: number;
   invalidDerivations?: number;
   duplicateDerivations?: number;
   duplicateSameParentDerivations?: number;
-  nullDerivationsByStrategy?: {[index: string]: number};
-  invalidDerivationsByStrategy?: {[index: string]: number};
-  duplicateDerivationsByStrategy?: {[index: string]: number};
-  duplicateSameParentDerivationsByStrategy?: {[index: string]: number};
+  nullDerivationsByStrategy?: Dictionary<number>;
```
[the remark this answers is not in the document]
Comment (Cypher1) [reply src/planning/strategizer.ts]: I think index or key would be a better name but most (if not all) of typescript deals with this transparently.

I think we could have problems with .toString (but that shouldn't be used on functions anyway).

File: src/runtime/particle-execution-context.ts [hunk 5]
```diff
@@ -39,7 +40,7 @@ export class ParticleExecutionContext {
   private readonly loader: Loader;
   private readonly pendingLoads = <Promise<void>[]>[];
   private readonly scheduler: StorageProxyScheduler = new StorageProxyScheduler();
-  private readonly keyedProxies: { [index: string]: StorageProxy | Promise<StorageProxy>} = {};
+  private readonly keyedProxies: Dictionary<StorageProxy | Promise<StorageProxy>> = {};
```
[the remark this answers is not in the document]
Comment (Cypher1) [reply src/runtime/particle-execution-context.ts]: I agree that that may be useful, the name MaybeAsync, to me, implies a Maybe<Async<T>> (i.e. Just (Async T) or Nothing) which isn't what we desire.

Maybe something like AnySync (as opposed to Async)? Still, bikeshedding can be delayed until we see that there is a strong need.

File: src/runtime/recipe/comparable.ts [hunk 6]
```diff
@@ -38,7 +39,7 @@ export function compareArrays<a>(a1: a[], a2: a[], compare: (first: a, second: a
...
   const keys = Object.keys(o1);
   let result: number;
   if ((result = compareNumbers(keys.length, Object.keys(o2).length)) !== 0) return result;
```
[the remark this answers is not in the document]
Comment (Cypher1) [reply src/runtime/recipe/comparable.ts]: 2nded. Is this something you have time for? Would be happy to explore it later but will have to delay it.

File: src/runtime/recipe/recipe.ts [hunk 7]
```diff
@@ -523,7 +524,7 @@ export class Recipe implements Cloneable<Recipe> {
   }
   
   // tslint:disable-next-line: no-any
-  updateToClone(dict): {[index: string]: any} {
+  updateToClone(dict: Dictionary<any>): Dictionary<any> {
```
[the remark this answers is not in the document]
Comment (Cypher1) [reply src/runtime/recipe/recipe.ts]: I think we should, but as a separate change. I haven't tested it with unknown.

File: src/runtime/services.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Cypher1) [reply src/runtime/services.ts]: Sounds good, but I suggest that when it becomes an interface again, we should make it

```
export interface Service extends Dictionary<Function {
...
}
```
and so this change would stand.

File: src/runtime/storage/storage-provider-base.ts (hunk 1 above)
Comment (Cypher1) [reply src/runtime/storage/storage-provider-base.ts]: Do you mean to rename Callback to KwargCallback or to create a specialization?

The request is unclear to me.

File: src/runtime/recipe/comparable.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Cypher1) [reply src/runtime/recipe/comparable.ts]: I think it is 'correctish', as in, it has the same runtime semantics, but different type semantics (not that they make a difference in this case).

I think we should examine actually enforcing that the compared objects have the same keys (using the Record type), but looking at the compareObjects code, we actually do need to support differently keyed objects (though the code that runs in that case is fairly trivial).

**Pull request merged**